### PR TITLE
ADAL hotfix 6.0.3

### PR DESCRIPTION
**Pull request by hieunguyenmsft**
## Proposed changes

add more logs in cache query in ADAL to investigate Outlook + MAM SDK issue (hot fix 6.0.3)

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [x] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m [hunk 1]
```diff
@@ -147,8 +148,8 @@ - (nullable MSIDCredentialCacheItem *)getCredential:(nonnull MSIDDefaultCredenti
 {
     assert(cacheQuery);
 
-    MSID_LOG_VERBOSE(context, @"(Default cache) Get accounts with environment %@", cacheQuery.environment);
-    MSID_LOG_VERBOSE_PII(context, @"(Default cache) Get accounts with environment %@, unique user id %@", cacheQuery.environment, cacheQuery.homeAccountId);
+    MSID_LOG_VERBOSE(context, @"(Default account cache) Get accounts with environment %@, hashed homeAccountId %@, isExactMatch %@, environment %@, environmentAliases %@", cacheQuery.environment, cacheQuery.homeAccountId, cacheQuery.exactMatch?@"yes":@"no" , cacheQuery.environment, cacheQuery.environmentAliases);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m:151]: cacheQuery.homeAccountId - this not actually hashed accountId, right?

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 2]
```diff
@@ -308,7 +309,7 @@ - (MSIDAccount *)accountForIdentifier:(MSIDAccountIdentifier *)accountIdentifier
                               context:(id<MSIDRequestContext>)context
                                 error:(NSError **)error
 {
-    MSID_LOG_VERBOSE(context, @"(Default accessor) Looking for account with client ID %@, family ID %@, authority %@", configuration.clientId, familyId, configuration.authority);
+    MSID_LOG_VERBOSE(context, @"(Default accessor) Looking for account with client ID %@, family ID %@, authority %@, hashed legacyAccountid %@, hashed homeAccountId %@", configuration.clientId, familyId, configuration.authority, [accountIdentifier.legacyAccountId msidTokenHash], [accountIdentifier.homeAccountId msidTokenHash]);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m:312]: Can you store pre-calculated hashes in the MSIDAccountIdentifier object? I believe we do a similar thing in latest common core. Otherwise, we'd do a lot of hash calculation every time when something is logged.

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 3]
```diff
@@ -323,7 +324,13 @@ - (MSIDAccount *)accountForIdentifier:(MSIDAccountIdentifier *)accountIdentifier
         MSID_LOG_WARN(context, @"(Default accessor) Failed to retrieve account with client ID %@, family ID %@, authority %@", configuration.clientId, familyId, configuration.authority);
         return nil;
     }
-
+    
+    MSID_LOG_VERBOSE(context, @"(Default accessor) number of accounts from default cache %lu", (unsigned long)accountCacheItems.count);
+    for (MSIDAccountCacheItem *cacheItem in accountCacheItems)
+    {
+        MSID_LOG_VERBOSE(context, @"(Default accessor) cacheItem with hashed username %@, homeAccountId %@, localAccountid %@, alternativeAccountId %@", cacheItem.username, cacheItem.homeAccountId, cacheItem.localAccountId, cacheItem.alternativeAccountId);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m:331]: Are those actually hashed?

**Review by oldalton — CHANGES_REQUESTED** [review]
Added a few comments/suggestions

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m (hunk 2 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: addressed. Just to clarify: I looked up at the code in common-core using by MSAL and I see the hashed property there. But I believe every time we call that property, the hash is calculated so no cache. I added a temp cache in this version of code, please have a look (it will be wrong if the value of id changed, but I don't think we are really re-assigning that property.)

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m (hunk 3 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: addressed

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m [hunk 4]
```diff
@@ -157,14 +158,18 @@ - (nullable MSIDCredentialCacheItem *)getCredential:(nonnull MSIDDefaultCredenti
         NSMutableArray<MSIDAccountCacheItem *> *filteredResults = [NSMutableArray array];
 
         BOOL shouldMatchAccount = !cacheQuery.homeAccountId || !cacheQuery.environment;
+        MSID_LOG_VERBOSE(context, @"(Default account cache) shouldMatchAccount %@", shouldMatchAccount? @"yes" : @"no");
 
         for (MSIDAccountCacheItem *cacheItem in cacheItems)
         {
+            MSID_LOG_VERBOSE(context, @"(Default account cache) cacheItem hashed home acccount ID %@, environment %@", [cacheItem.homeAccountId msidTokenHash], cacheItem.environment);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m:165]: Discussed offline and agreed to add caching for cacheItem has as well

Comment (hieunguyenmsft) [reply IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m]: Addressed

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m [hunk 5]
```diff
@@ -147,8 +148,8 @@ - (nullable MSIDCredentialCacheItem *)getCredential:(nonnull MSIDDefaultCredenti
 {
     assert(cacheQuery);
 
-    MSID_LOG_VERBOSE(context, @"(Default cache) Get accounts with environment %@", cacheQuery.environment);
-    MSID_LOG_VERBOSE_PII(context, @"(Default cache) Get accounts with environment %@, unique user id %@", cacheQuery.environment, cacheQuery.homeAccountId);
+    MSID_LOG_VERBOSE(context, @"(Default account cache) Get accounts with environment %@, hashed homeAccountId %@, isExactMatch %@, environment %@, environmentAliases %@", cacheQuery.environment, [cacheQuery.homeAccountId msidTokenHash], cacheQuery.exactMatch?@"yes":@"no" , cacheQuery.environment, cacheQuery.environmentAliases);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m:151]: Doesn't cacheQuery inherit from cacheItem? if yes, you can use hashedHomeAccountId instead of [cacheQuery.homeAccountId msidTokenHash]

File: IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheAccessor.m [hunk 6]
```diff
@@ -738,7 +741,7 @@ - (MSIDBaseToken *)getTokenByLegacyUserId:(NSString *)legacyUserId
 
     for (NSURL *alias in aliases)
     {
-        MSID_LOG_VERBOSE(context, @"(Legacy accessor) Looking for token with alias %@, clientId %@, resource %@", alias, clientId, resource);
+        MSID_LOG_VERBOSE(context, @"(Legacy accessor) Looking for token with alias %@, clientId %@, resource %@, hashed legacy userid %@", alias, clientId, resource, [legacyUserId msidTokenHash]);
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheAccessor.m:744]: You can call [legacyUserId msidTokenHash] once on line 741 before going into the aliases loop. It should be the same for all aliases.

File: IdentityCore/src/cache/token/MSIDAccountCacheItem.m [hunk 7]
```diff
@@ -32,6 +32,11 @@ @interface MSIDAccountCacheItem()
 
 @implementation MSIDAccountCacheItem
 
+@synthesize hashedHomeAccountId = _hashedHomeAccountId;
```
Comment (oldalton) [inline IdentityCore/src/cache/token/MSIDAccountCacheItem.m:35]: Why is this needed?

File: IdentityCore/src/oauth2/account/MSIDAccountIdentifier.m [hunk 8]
```diff
@@ -23,9 +23,12 @@
 
 #import "MSIDAccountIdentifier.h"
 #import "MSIDClientInfo.h"
+#import "NSString+MSIDExtensions.h"
 
 @implementation MSIDAccountIdentifier
 
+@synthesize hashedHomeAccountId = _hashedHomeAccountId;
```
Comment (oldalton) [inline IdentityCore/src/oauth2/account/MSIDAccountIdentifier.m:30]: Same here - why is it needed? Can you use readwrite properties instead declared in .m?

**Review by oldalton — COMMENTED** [review]
Added a few more suggestions

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m (hunk 5 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m:151]: Hmm I checked and these 2 classes are from different families.

File: IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheAccessor.m (hunk 6 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/cache/accessor/MSIDLegacyTokenCacheAccessor.m:744]: addressed

File: IdentityCore/src/oauth2/account/MSIDAccountIdentifier.m (hunk 8 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/oauth2/account/MSIDAccountIdentifier.m:30]: Wouldn't readwrite allow the value is assigned externally? IMO I think we shouldn't let the hashValue assigned outside, it should be calculated inside the class. But if that's ok, I will update these properties to readwrite.
Using @synthesize is for assigning the ivar value is the only way I know. If I set the property @readonly I can't use _ivar without that synthesize declaration.

Comment (hieunguyenmsft) [reply IdentityCore/src/oauth2/account/MSIDAccountIdentifier.m]: Addressed as discussed offline.

**Review by oldalton — APPROVED** [review]
